core/commands/storage/session.go

```diff
@@ -346,8 +346,11 @@ func (ss *FileContracts) IncrementContract(shard *Shard, signedEscrowContract []
 	ss.Lock()
 	defer ss.Unlock()
 
-	if ss.GuardContracts == nil {
-		ss.GuardContracts = make([]*guardpb.Contract, len(ss.ShardInfo))
+	// expand guard contracts storage according to number of shards
+	if shard.ShardIndex >= len(ss.GuardContracts) {
```

Comment:
how ShardIndex become == ss.GuardContracts? My understanding is the ShardIndex should be < ss.GuardContracts

Reply:
> how ShardIndex become == ss.GuardContracts? My understanding is the ShardIndex should be < ss.GuardContracts

The length of GuardContracts slice start from 0, and can be increased automatically.

Reply:
This logic is definitely not ideal - guard contracts are gradually obtained but it should really be set the length during init - instead of waiting for increasing shard index. Should get re-factored in the future.